src/cli/src/profile/sync/cloud/login_agent.rs

```diff
@@ -0,0 +1,255 @@
+    pub async fn download_profile(&mut self) -> Result<(), CloudError> {
+        // Check whether we have credentials in session or on disk
+        let creds = match self.session.as_ref() {
+            // First, try to get the token from the agent session
+            Some(creds) => creds,
+            None => {
+                // If that doesn't work, try to get the token from disk
+                let loaded_creds = Credentials::try_load(&self.path).await?;
+                self.session.replace(loaded_creds);
+                self.session.as_ref().unwrap()
+            }
+        };
+
+        let _result = self.try_download_profile(creds).await?;
+        Ok(())
+    }
+
+    /// Attempts to download a Fluvio Cloud profile with the given credentials
+    async fn try_download_profile(&self, creds: &Credentials) -> Result<(), CloudError> {
+        let response = surf::get(format!("{}/api/v1/downloadProfile", &self.remote))
+            .set_header("Authorization", &*creds.token)
+            .await?;
+        match response.status() {
+            StatusCode::Ok => {
+                info!("Successfully authenticated with token!");
+                error!("TODO, implement profile download");
+                Ok(())
+            }
+            StatusCode::Unauthorized => {
+                warn!("Failed to download profile, token is expired or invalid");
+                Err(CloudError::Unauthorized)
+            }
+            StatusCode::NotFound => {
+                warn!("Failed to download a profile for this user");
+                Err(CloudError::ProfileNotFound)
+            }
+            _ => Err(CloudError::Unexpected),
+        }
+    }
+
+    /// Attempts to use a username and password to login to Fluvio Cloud.
+    ///
+    /// If this succeeds, the LoginAgent will save the Fluvio Cloud
+    /// credentials in a session to be used later.
+    pub async fn authenticate(
+        &mut self,
+        email: String,
+        password: String,
+    ) -> Result<(), CloudError> {
+        let login = LoginRequest {
+            email: email.clone(),
+            password,
+        };
+        let mut response = surf::post(format!("{}/api/v1/loginUser", &self.remote))
+            .body_json(&login)
+            .unwrap()
+            .await?;
+
+        match response.status() {
+            StatusCode::Ok => {
+                let creds = response.body_json::<Credentials>().await?;
+                self.save_credentials(creds).await?;
+                Ok(())
+            }
+            _ => {
+                warn!("Failed to login");
+                Err(CloudError::AuthenticationError(email))
+            }
+        }
+    }
+
+    async fn save_credentials(&mut self, creds: Credentials) -> Result<(), CloudError> {
+        // Save credentials in agent session
+        self.session.replace(creds);
+        // Save credentials to disk
+        self.session.as_ref().unwrap().try_save(&self.path).await?;
```

Comment:
This could be better if save is done first then replace it in memory:
- it is more clear
- avoid saving credential in memory if failure happens